### ISSUE TITLE
OpProfiler wish list

### ISSUE THREAD

**Issue by raver119** [deeplearning4j/nd4j#2579]
This issue is compilation of desired additional functionality for OpProfiler:
[ ] - NAN_PANIC/INF_PANIC/ANY_PANIC should have some form of flexible configuration, so user should be able to choose, what he's going to track. This will also allow to reduce perf penalty coming from additional op calls.
[ ] - SCOPE_PANIC should provide additional checks for CIRCULAR workspaces (as used in AsyncDataSetIterators), to ensure there's no buffer overruns.

**Comment by AlexDBlack** [comment]
A couple of other random ideas - things we *could* do, though I'm not sure how useful these would be in practice:
- re: flexible config - filtering based on workspace or thread (i.e., profile only some of these)
- Panic for conditions not based on INDArray contents:
    - Total memory use
    - Array size/length exceeds X

**Comment by raver119** [comment]
Re: targeting specific workspace/thread - sounds interesting.
Re: conditions. Total memory use of? Workspace? NDArray? Any of those?

**Comment by AlexDBlack** [comment]
Yeah, memory use of worspace or single INDArray might be useful at times.
For example, exceeding say 1gb total allocated in your workspace might indicate an error (for example, on a small network or something that shouldn't use much memory). Or a single array exceeding some manually calculated maximum size.
